Avoid fast-path in --sh-boot script when PEX_TOOLS=1

huonw opened the pull request:
This patch stops the "execute venv entrypoint directly" fast-path from happening when the user has set `PEX_TOOLS=1` with a `--sh-boot` PEX, mimicking the similar check performed for Python bootstrapping:

https://github.com/pex-tool/pex/blob/17bd416647eaca978e538408f1ec21813ab26335/pex/pex_boot.py#L204-L205

This change ensures that `PEX_TOOLS=1 ./some.pex` always works, no matter the state of the `PEX_ROOT` cache.

Fixes #2725

huonw commented:
I haven't added a test yet, to confirm the direction: is the sh-boot script the right place to put this, or is it better placed in `venv_pex.py` somehow?

jsirois commented:
> I haven't added a test yet, to confirm the direction: is the sh-boot script the right place to put this, or is it better placed in `venv_pex.py` somehow?

This is the right place. The venv does not contain any PEX `.bootstrap/` code which is where the pex tools code lives; so the `venv_pex.py` code would need to re-exec back into the `PEX` exported by the sh-boot script. You'd have:
```
[PEX_TOOLS=1 PEX(sh-boot) ...] -> # Original PEX file invocation.
    <- exec `venv/bin/python venv/ ...` # The sh-boot script re-direction to existing venv.
[venv/bin/python venv/__main__.py] -> # The venv invocation.
    exec `PEX_TOOLS=1 sys.executable PEX ...` # The venv script re-direction to the original PEX file.
```
So that's 2 Python binary startups.

With your change, you just have 1 Python binary startup.

My shortcut accounting for a Python invocation with some stdlib imports at least is ~50ms.

jsirois commented:
@huonw if you want to update `pex/version.py` to 2.33.7 for this bug fix and add an entry in `CHANGES.md` I can get out a release today and save some trees.

huonw commented on pex/sh_boot.py:
```diff
@@ -199,10 +199,11 @@ def create_sh_boot_script(
         INSTALLED_PEX="${{PEX_ROOT}}/{pex_installed_relpath}"
 
-        if [ -n "${{VENV}}" -a -x "${{INSTALLED_PEX}}" ]; then
+        if [ -n "${{VENV}}" -a -x "${{INSTALLED_PEX}}" -a "${{PEX_TOOLS:-}}" != "" ]; then
```
Oh, yes 🤦

`-z` is much better. Thanks.

huonw commented:
I've updated:

- using `-z`
- added a test
- added the release metadata

Thanks for the initial review.

> The venv does not contain any PEX .bootstrap/ code which is where the pex tools code lives

Ah, yes, that makes the "proper" approach (that in this PR) clearly the better one. Thank you.

jsirois commented:
Thanks @huonw. This is now available in 2.33.7: https://github.com/pex-tool/pex/releases/tag/v2.33.7
